File: backend/services/signal_engine.py
import pandas as pd
from typing import Optional
from config import (
    ADX_STRONG_TREND, ADX_MODERATE_TREND,
    WEIGHT_MOMENTUM, WEIGHT_TREND, WEIGHT_VOLATILITY, WEIGHT_VOLUME,
)
from models.signal import ConsolidatedSignal, MonthlyTrendSignal

Signal = str  # "Buy", "Sell", "Neutral"


class SignalEngine:
    """Compute individual indicator signals and consolidated recommendation."""

    # --- Individual Signal Rules ---

    def signal_sma_cross(self, indicators: dict) -> Signal:
        """Golden Cross / Death Cross: SMA 50 vs SMA 200."""
        sma_50 = indicators.get("sma_50")
        sma_200 = indicators.get("sma_200")
        if sma_50 is None or sma_200 is None:
            return "Neutral"
        s50 = _last_valid(sma_50)
        s200 = _last_valid(sma_200)
        if s50 is None or s200 is None:
            return "Neutral"
        if s50 > s200:
            return "Buy"
        elif s50 < s200:
            return "Sell"
        return "Neutral"

    def signal_ema(self, indicators: dict, close_price: float) -> Signal:
        """Price vs EMA: price > EMA = Buy."""
        ema_12 = _last_valid(indicators.get("ema_12"))
        if ema_12 is None:
            return "Neutral"
        if close_price > ema_12:
            return "Buy"
        elif close_price < ema_12:
            return "Sell"
        return "Neutral"

    def signal_rsi(self, indicators: dict) -> Signal:
        """RSI: <30 = Buy (oversold), >70 = Sell (overbought)."""
        rsi = _last_valid(indicators.get("rsi"))
        if rsi is None:
            return "Neutral"
        if rsi < 30:
            return "Buy"
        elif rsi > 70:
            return "Sell"
        return "Neutral"

    def signal_macd(self, indicators: dict) -> Signal:
        """MACD line > Signal line = Buy."""
        macd_data = indicators.get("macd", {})
        macd_line = _last_valid(macd_data.get("macd"))
        signal_line = _last_valid(macd_data.get("signal"))
        if macd_line is None or signal_line is None:
            return "Neutral"
        if macd_line > signal_line:
            return "Buy"
        elif macd_line < signal_line:
            return "Sell"
        return "Neutral"

    def signal_bollinger(self, indicators: dict, close_price: float) -> Signal:
        """Price below lower band = Buy, above upper = Sell."""
        bb = indicators.get("bollinger", {})
        lower = _last_valid(bb.get("lower"))
        upper = _last_valid(bb.get("upper"))
        if lower is None or upper is None:
            return "Neutral"
        if close_price < lower:
            return "Buy"
        elif close_price > upper:
            return "Sell"
        return "Neutral"

    def signal_williams_r(self, indicators: dict) -> Signal:
        """Williams %R: < -80 = Buy, > -20 = Sell."""
        wr = _last_valid(indicators.get("williams_r"))
        if wr is None:
            return "Neutral"
        if wr < -80:
            return "Buy"
        elif wr > -20:
            return "Sell"
        return "Neutral"

    def signal_mfi(self, indicators: dict) -> Signal:
        """MFI: <20 = Buy, >80 = Sell."""
        mfi = _last_valid(indicators.get("mfi"))
        if mfi is None:
            return "Neutral"
        if mfi < 20:
            return "Buy"
        elif mfi > 80:
            return "Sell"
        return "Neutral"

    def signal_roc(self, indicators: dict) -> Signal:
        """ROC: > 0 and rising = Buy confirmation, < 0 and falling = Sell."""
        roc = indicators.get("roc")
        if roc is None:
            return "Neutral"
        roc_val = _last_valid(roc)
        if roc_val is None:
            return "Neutral"
        if roc_val > 0:
            return "Buy"
        elif roc_val < 0:
            return "Sell"
        return "Neutral"

    def signal_vwap(self, indicators: dict, close_price: float) -> Signal:
        """Price > VWAP = Buy, < VWAP = Sell."""
        vwap = _last_valid(indicators.get("vwap"))
        if vwap is None:
            return "Neutral"
        if close_price > vwap:
            return "Buy"
        elif close_price < vwap:
            return "Sell"
        return "Neutral"

    # --- Consolidated Signal ---

    def compute_all_signals(
        self, indicators: dict, close_price: float,
    ) -> dict[str, Signal]:
        """Compute signals for all indicators."""
        return {
            "SMA Cross": self.signal_sma_cross(indicators),
            "EMA": self.signal_ema(indicators, close_price),
            "RSI": self.signal_rsi(indicators),
            "MACD": self.signal_macd(indicators),
            "Bollinger Bands": self.signal_bollinger(indicators, close_price),
            "Williams %R": self.signal_williams_r(indicators),
            "MFI": self.signal_mfi(indicators),
            "ROC": self.signal_roc(indicators),
            "VWAP": self.signal_vwap(indicators, close_price),
        }

    def compute_consolidated(
        self, indicators: dict, close_price: float,
    ) -> ConsolidatedSignal:
        signals = self.compute_all_signals(indicators, close_price)
        adx_val = _last_valid(indicators.get("adx"))

        # Assign numeric values
        def sig_val(s: str) -> float:
            return {"Buy": 1.0, "Sell": -1.0, "Neutral": 0.0}.get(s, 0.0)

        # Weighted categories
        momentum_sigs = [signals["Williams %R"], signals["RSI"], signals["MFI"]]
        trend_sigs = [signals["SMA Cross"], signals["EMA"], signals["MACD"]]
        volatility_sigs = [signals["Bollinger Bands"]]
        volume_sigs = [signals["ROC"]]
        if signals.get("VWAP") != "Neutral":
            volume_sigs.append(signals["VWAP"])

        def avg_signal(sigs: list[str]) -> float:
            if not sigs:
                return 0.0
            return sum(sig_val(s) for s in sigs) / len(sigs)

        score = (
            WEIGHT_MOMENTUM * avg_signal(momentum_sigs)
            + WEIGHT_TREND * avg_signal(trend_sigs)
            + WEIGHT_VOLATILITY * avg_signal(volatility_sigs)
            + WEIGHT_VOLUME * avg_signal(volume_sigs)
        )

        # Map score to level
        if score >= 0.6:
            level = "Strong Buy"
        elif score >= 0.2:
            level = "Buy"
        elif score > -0.2:
            level = "Hold"
        elif score > -0.6:
            level = "Sell"
        else:
            level = "Strong Sell"

        # ADX confidence
        if adx_val is not None and adx_val > ADX_STRONG_TREND:
            adx_confidence = "high"
        elif adx_val is not None and adx_val > ADX_MODERATE_TREND:
            adx_confidence = "moderate"
        else:
            adx_confidence = "low"

        buy_count = sum(1 for s in signals.values() if s == "Buy")
        sell_count = sum(1 for s in signals.values() if s == "Sell")
        neutral_count = sum(1 for s in signals.values() if s == "Neutral")

        explanation = self._generate_explanation(
            signals, adx_val, adx_confidence, level, score,
        )

        return ConsolidatedSignal(
            signal=level,
            score=round(score, 3),
            explanation=explanation,
            adx_value=round(adx_val, 2) if adx_val is not None else None,
            adx_confidence=adx_confidence,
            individual_signals=signals,
            buy_count=buy_count,
            sell_count=sell_count,
            neutral_count=neutral_count,
        )

    def _generate_explanation(
        self,
        signals: dict[str, str],
        adx_val: Optional[float],
        adx_confidence: str,
        level: str,
        score: float,
    ) -> str:
        buy_indicators = [k for k, v in signals.items() if v == "Buy"]
        sell_indicators = [k for k, v in signals.items() if v == "Sell"]
        neutral_indicators = [k for k, v in signals.items() if v == "Neutral"]

        parts = [f"{level}"]

        if adx_val is not None:
            adx_label = {
                "high": "Strong trend",
                "moderate": "Moderate trend",
                "low": "Weak/no trend",
            }[adx_confidence]
            parts.append(f"(ADX: {adx_val:.0f} — {adx_label})")

        total = len(signals)
        parts.append(
            f"— {len(buy_indicators)} of {total} indicators signal Buy."
        )

        if buy_indicators:
            parts.append(f"Bullish: {', '.join(buy_indicators)}.")
        if sell_indicators:
            parts.append(f"Bearish: {', '.join(sell_indicators)}.")

        if adx_confidence == "low":
            parts.append(
                "Note: The market is not trending; signals may be less reliable."
            )

        return " ".join(parts)


def _last_valid(series) -> Optional[float]:
    """Get the last non-NaN value from a pandas Series or return None."""
    if series is None:
        return None
    if isinstance(series, (int, float)):
        return float(series)
    if isinstance(series, pd.Series):
        valid = series.dropna()
        if valid.empty:
            return None
        return float(valid.iloc[-1])
    return None

File: frontend/src/lib/services/signalEngine.ts
/**
 * Signal engine for computing individual indicator signals and consolidated recommendations.
 */

import {
  WEIGHT_MOMENTUM,
  WEIGHT_TREND,
  WEIGHT_VOLATILITY,
  WEIGHT_VOLUME,
  ADX_STRONG_TREND,
  ADX_MODERATE_TREND,
} from "../config";

import type { IndicatorResults } from "./indicatorCalculator";
import type { IndicatorSignal, ConsolidatedSignalLevel, ConsolidatedSignal } from "@/types";

// Get the last valid value from an array
function lastValid(arr: (number | null | undefined)[]): number | null {
  if (!arr || arr.length === 0) return null;
  for (let i = arr.length - 1; i >= 0; i--) {
    const val = arr[i];
    if (val !== null && val !== undefined && !isNaN(val)) {
      return val;
    }
  }
  return null;
}

// Individual signal functions
export function signalSMACross(indicators: IndicatorResults): IndicatorSignal {
  const sma50 = lastValid(indicators.sma_50);
  const sma200 = lastValid(indicators.sma_200);

  if (sma50 === null || sma200 === null) {
    return "Neutral";
  }

  if (sma50 > sma200) {
    return "Buy";
  } else if (sma50 < sma200) {
    return "Sell";
  }
  return "Neutral";
}

export function signalEMA(indicators: IndicatorResults, closePrice: number): IndicatorSignal {
  const ema12 = lastValid(indicators.ema_12);

  if (ema12 === null) {
    return "Neutral";
  }

  if (closePrice > ema12) {
    return "Buy";
  } else if (closePrice < ema12) {
    return "Sell";
  }
  return "Neutral";
}

export function signalRSI(indicators: IndicatorResults): IndicatorSignal {
  const rsi = lastValid(indicators.rsi);

  if (rsi === null) {
    return "Neutral";
  }

  if (rsi < 30) {
    return "Buy";
  } else if (rsi > 70) {
    return "Sell";
  }
  return "Neutral";
}

export function signalMACD(indicators: IndicatorResults): IndicatorSignal {
  const macdLine = lastValid(indicators.macd.macd);
  const signalLine = lastValid(indicators.macd.signal);

  if (macdLine === null || signalLine === null) {
    return "Neutral";
  }

  if (macdLine > signalLine) {
    return "Buy";
  } else if (macdLine < signalLine) {
    return "Sell";
  }
  return "Neutral";
}

export function signalBollinger(indicators: IndicatorResults, closePrice: number): IndicatorSignal {
  const lower = lastValid(indicators.bollinger.lower);
  const upper = lastValid(indicators.bollinger.upper);

  if (lower === null || upper === null) {
    return "Neutral";
  }

  if (closePrice < lower) {
    return "Buy";
  } else if (closePrice > upper) {
    return "Sell";
  }
  return "Neutral";
}

export function signalWilliamsR(indicators: IndicatorResults): IndicatorSignal {
  const wr = lastValid(indicators.williams_r);

  if (wr === null) {
    return "Neutral";
  }

  if (wr < -80) {
    return "Buy";
  } else if (wr > -20) {
    return "Sell";
  }
  return "Neutral";
}

export function signalMFI(indicators: IndicatorResults): IndicatorSignal {
  const mfi = lastValid(indicators.mfi);

  if (mfi === null) {
    return "Neutral";
  }

  if (mfi < 20) {
    return "Buy";
  } else if (mfi > 80) {
    return "Sell";
  }
  return "Neutral";
}

export function signalROC(indicators: IndicatorResults): IndicatorSignal {
  const roc = lastValid(indicators.roc);

  if (roc === null) {
    return "Neutral";
  }

  if (roc > 0) {
    return "Buy";
  } else if (roc < 0) {
    return "Sell";
  }
  return "Neutral";
}

// Compute all individual signals
export function computeAllSignals(
  indicators: IndicatorResults,
  closePrice: number
): Record<string, IndicatorSignal> {
  return {
    "SMA Cross": signalSMACross(indicators),
    EMA: signalEMA(indicators, closePrice),
    RSI: signalRSI(indicators),
    MACD: signalMACD(indicators),
    "Bollinger Bands": signalBollinger(indicators, closePrice),
    "Williams %R": signalWilliamsR(indicators),
    MFI: signalMFI(indicators),
    ROC: signalROC(indicators),
  };
}

// Signal value mapping
function sigVal(signal: IndicatorSignal): number {
  switch (signal) {
    case "Buy":
      return 1.0;
    case "Sell":
      return -1.0;
    default:
      return 0.0;
  }
}

// Average of signals
function avgSignal(signals: IndicatorSignal[]): number {
  if (signals.length === 0) return 0.0;
  const sum = signals.reduce((acc, s) => acc + sigVal(s), 0);
  return sum / signals.length;
}

// Map score to signal level
function scoreToLevel(score: number): ConsolidatedSignalLevel {
  if (score >= 0.6) return "Strong Buy";
  if (score >= 0.2) return "Buy";
  if (score > -0.2) return "Hold";
  if (score > -0.6) return "Sell";
  return "Strong Sell";
}

// Get ADX confidence
function getADXConfidence(adxValue: number | null): "high" | "moderate" | "low" {
  if (adxValue === null) return "low";
  if (adxValue > ADX_STRONG_TREND) return "high";
  if (adxValue > ADX_MODERATE_TREND) return "moderate";
  return "low";
}

// Generate explanation
function generateExplanation(
  signals: Record<string, IndicatorSignal>,
  adxValue: number | null,
  adxConfidence: "high" | "moderate" | "low",
  level: ConsolidatedSignalLevel,
  score: number
): string {
  const buyIndicators = Object.entries(signals)
    .filter(([, v]) => v === "Buy")
    .map(([k]) => k);
  const sellIndicators = Object.entries(signals)
    .filter(([, v]) => v === "Sell")
    .map(([k]) => k);

  const parts: string[] = [level];

  if (adxValue !== null) {
    const adxLabel = {
      high: "Strong trend",
      moderate: "Moderate trend",
      low: "Weak/no trend",
    }[adxConfidence];
    parts.push(`(ADX: ${adxValue.toFixed(0)} — ${adxLabel})`);
  }

  const total = Object.keys(signals).length;
  parts.push(`— ${buyIndicators.length} of ${total} indicators signal Buy.`);

  if (buyIndicators.length > 0) {
    parts.push(`Bullish: ${buyIndicators.join(", ")}.`);
  }
  if (sellIndicators.length > 0) {
    parts.push(`Bearish: ${sellIndicators.join(", ")}.`);
  }

  if (adxConfidence === "low") {
    parts.push("Note: The market is not trending; signals may be less reliable.");
  }

  return parts.join(" ");
}

// Compute consolidated signal
export function computeConsolidatedSignal(
  indicators: IndicatorResults,
  closePrice: number
): ConsolidatedSignal {
  const signals = computeAllSignals(indicators, closePrice);
  const adxValue = lastValid(indicators.adx);

  // Weighted categories
  const momentumSigs: IndicatorSignal[] = [
    signals["Williams %R"],
    signals.RSI,
    signals.MFI,
  ];
  const trendSigs: IndicatorSignal[] = [
    signals["SMA Cross"],
    signals.EMA,
    signals.MACD,
  ];
  const volatilitySigs: IndicatorSignal[] = [signals["Bollinger Bands"]];
  const volumeSigs: IndicatorSignal[] = [signals.ROC];

  // Calculate weighted score
  const score =
    WEIGHT_MOMENTUM * avgSignal(momentumSigs) +
    WEIGHT_TREND * avgSignal(trendSigs) +
    WEIGHT_VOLATILITY * avgSignal(volatilitySigs) +
    WEIGHT_VOLUME * avgSignal(volumeSigs);

  const level = scoreToLevel(score);
  const adxConfidence = getADXConfidence(adxValue);

  const buyCount = Object.values(signals).filter((s) => s === "Buy").length;
  const sellCount = Object.values(signals).filter((s) => s === "Sell").length;
  const neutralCount = Object.values(signals).filter((s) => s === "Neutral").length;

  const explanation = generateExplanation(signals, adxValue, adxConfidence, level, score);

  return {
    signal: level,
    score: Math.round(score * 1000) / 1000,
    explanation,
    adx_value: adxValue !== null ? Math.round(adxValue * 100) / 100 : null,
    adx_confidence: adxConfidence,
    individual_signals: signals,
    buy_count: buyCount,
    sell_count: sellCount,
    neutral_count: neutralCount,
  };
}

// Get signal explanation for individual indicators
export function getSignalExplanation(
  name: string,
  signal: IndicatorSignal,
  currentValue: number | null
): string {
  if (currentValue === null) {
    return "Insufficient data to compute this indicator.";
  }

  const explanations: Record<string, Record<IndicatorSignal, string>> = {
    rsi: {
      Buy: `RSI is at ${currentValue.toFixed(1)} — the stock is in oversold territory, suggesting it may be due for a bounce.`,
      Sell: `RSI is at ${currentValue.toFixed(1)} — the stock is in overbought territory, suggesting it may be due for a pullback.`,
      Neutral: `RSI is at ${currentValue.toFixed(1)} — momentum is neutral, no strong signal.`,
    },
    williams_r: {
      Buy: `Williams %R is at ${currentValue.toFixed(1)} — near the bottom of its recent range (oversold).`,
      Sell: `Williams %R is at ${currentValue.toFixed(1)} — near the top of its recent range (overbought).`,
      Neutral: `Williams %R is at ${currentValue.toFixed(1)} — mid-range, no strong signal.`,
    },
    mfi: {
      Buy: `MFI is at ${currentValue.toFixed(1)} — strong selling pressure with high volume (oversold).`,
      Sell: `MFI is at ${currentValue.toFixed(1)} — strong buying pressure with high volume (overbought).`,
      Neutral: `MFI is at ${currentValue.toFixed(1)} — normal money flow, no extreme signal.`,
    },
    sma_cross: {
      Buy: "50-day SMA is above the 200-day SMA (Golden Cross) — bullish trend.",
      Sell: "50-day SMA is below the 200-day SMA (Death Cross) — bearish trend.",
      Neutral: "SMA lines are converging — trend direction is unclear.",
    },
    ema: {
      Buy: "Price is above the EMA — uptrend is intact.",
      Sell: "Price is below the EMA — downtrend signal.",
      Neutral: "Price is near the EMA — no clear direction.",
    },
    macd: {
      Buy: "MACD line is above the signal line — positive momentum.",
      Sell: "MACD line is below the signal line — negative momentum.",
      Neutral: "MACD and signal lines are converging — momentum is flat.",
    },
    bollinger: {
      Buy: "Price is near or below the lower Bollinger Band — potentially oversold.",
      Sell: "Price is near or above the upper Bollinger Band — potentially overbought.",
      Neutral: "Price is within the Bollinger Bands — normal range.",
    },
    roc: {
      Buy: `ROC is at ${currentValue.toFixed(2)}% — positive price momentum.`,
      Sell: `ROC is at ${currentValue.toFixed(2)}% — negative price momentum.`,
      Neutral: `ROC is near zero (${currentValue.toFixed(2)}%) — price is consolidating.`,
    },
  };

  const indicatorExplanations = explanations[name];
  if (indicatorExplanations && indicatorExplanations[signal]) {
    return indicatorExplanations[signal];
  }

  return `Current value: ${currentValue.toFixed(2)}`;
}
